File: deploy/deploy.js
#!/usr/bin/env node
/** Deploy PulseWatch: native Go binary + Next.js standalone + Nginx */
const { Client } = require('ssh2');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const { buildCorsOrigins } = require('./lib/cors-origins');

const HOST = process.env.DEPLOY_HOST || '49.234.112.108';
const USER = process.env.DEPLOY_USER || 'ubuntu';
const PASSWORD = process.env.DEPLOY_PASSWORD || 'prs@2018';
const APP_DIR = '/opt/pulsewatch';
const ROOT = path.join(__dirname, '..');

function exec(conn, cmd) {
  return new Promise((resolve) => {
    console.log(`>>> ${cmd.slice(0, 100)}`);
    conn.exec(cmd, (err, stream) => {
      if (err) { console.error(err); return resolve(1); }
      stream.on('data', (d) => process.stdout.write(d));
      stream.stderr.on('data', (d) => process.stderr.write(d));
      stream.on('close', (code) => resolve(code || 0));
    });
  });
}

function sudo(conn, cmd) {
  return exec(conn, `echo '${PASSWORD.replace(/'/g, "'\\''")}' | sudo -S bash -c ${JSON.stringify(cmd)}`);
}

function upload(sftp, local, remote) {
  return new Promise((res, rej) => sftp.fastPut(local, remote, (e) => e ? rej(e) : res()));
}

async function main() {
  // Cross-compile API
  console.log('Building Go API for linux/amd64...');
  execSync('go build -o pulsewatch-api ./cmd/server', {
    cwd: path.join(ROOT, 'apps/api'),
    env: { ...process.env, GOOS: 'linux', GOARCH: 'amd64', CGO_ENABLED: '0' },
    stdio: 'inherit',
  });

  const conn = new Client();
  await new Promise((res, rej) => {
    conn.on('ready', res);
    conn.on('error', rej);
    conn.connect({ host: HOST, port: 22, username: USER, password: PASSWORD, readyTimeout: 30000 });
  });
  console.log('SSH connected.');

  // Install deps
  await sudo(conn, 'apt-get update -qq && DEBIAN_FRONTEND=noninteractive apt-get install -y -qq nginx curl ca-certificates 2>/dev/null || true');
  await sudo(conn, 'systemctl enable nginx && systemctl start nginx || true');

  // Node.js 22 if missing
  await exec(conn, 'command -v node >/dev/null || (curl -fsSL https://deb.nodesource.com/setup_22.x | sudo -E bash - && sudo apt-get install -y nodejs)');

  // Database
  await sudo(conn, "sudo -u postgres psql -p 6541 -tc \"SELECT 1 FROM pg_database WHERE datname='pulsewatch'\" | grep -q 1 || sudo -u postgres psql -p 6541 -c \"CREATE DATABASE pulsewatch;\"");

  await sudo(conn, `mkdir -p ${APP_DIR}/{api,web} && chown -R ubuntu:ubuntu ${APP_DIR}`);

  const jwt1 = require('crypto').randomBytes(32).toString('hex');
  const jwt2 = require('crypto').randomBytes(32).toString('hex');
  const corsOrigins = buildCorsOrigins(HOST, process.env.APP_DOMAINS || '');
  const envContent = `DATABASE_URL=postgresql://postgres:prs%402018@127.0.0.1:6541/pulsewatch
JWT_SECRET=${jwt1}
JWT_REFRESH_SECRET=${jwt2}
PORT=4000
CORS_ORIGINS=${corsOrigins}
SMTP_MODE=console
NODE_ENV=production
HOSTNAME=0.0.0.0
`;

  const sftp = await new Promise((res, rej) => conn.sftp((e, s) => e ? rej(e) : res(s)));

  await upload(sftp, path.join(ROOT, 'apps/api/pulsewatch-api'), '/tmp/pulsewatch-api');
  await exec(conn, `mv /tmp/pulsewatch-api ${APP_DIR}/api/pulsewatch-api && chmod +x ${APP_DIR}/api/pulsewatch-api`);

  // Upload Next.js standalone
  const standalone = path.join(ROOT, 'apps/web/.next/standalone');
  const staticDir = path.join(ROOT, 'apps/web/.next/static');
  const publicDir = path.join(ROOT, 'apps/web/public');

  // Create web tarball locally
  const webTar = path.join(__dirname, 'web-bundle.tar.gz');
  if (fs.existsSync(standalone)) {
    execSync(`tar -czf "${webTar}" -C "${path.join(ROOT, 'apps/web')}" .next/standalone .next/static public`, { shell: true, stdio: 'inherit' });
    await upload(sftp, webTar, '/tmp/web-bundle.tar.gz');
    await exec(conn, `cd ${APP_DIR}/web && tar -xzf /tmp/web-bundle.tar.gz && rm /tmp/web-bundle.tar.gz`);
    await exec(conn, `cp -r ${APP_DIR}/web/.next/static ${APP_DIR}/web/.next/standalone/apps/web/.next/ && cp -r ${APP_DIR}/web/public ${APP_DIR}/web/.next/standalone/apps/web/`);
    fs.unlinkSync(webTar);
  } else {
    console.warn('Standalone build not found, uploading source for server build...');
    const srcTar = path.join(__dirname, 'web-src.tar.gz');
    execSync(`tar -czf "${srcTar}" --exclude=node_modules --exclude=.next -C "${path.join(ROOT, 'apps/web')}" .`, { shell: true });
    await upload(sftp, srcTar, '/tmp/web-src.tar.gz');
    await exec(conn, `cd ${APP_DIR}/web && tar -xzf /tmp/web-src.tar.gz && npm install && NEXT_PUBLIC_API_URL=http://${HOST} npm run build && cp -r .next/static .next/standalone/.next/ && cp -r public .next/standalone/`);
    fs.unlinkSync(srcTar);
  }

  await exec(conn, `cat > ${APP_DIR}/api/.env << 'EOF'\n${envContent}EOF`);
  await exec(conn, `cat > ${APP_DIR}/web/.env << 'EOF'\nINTERNAL_API_URL=http://127.0.0.1:4000\nPORT=3000\nHOSTNAME=0.0.0.0\nEOF`);

  // Systemd services
  const apiService = `[Unit]
Description=PulseWatch API
After=network.target

[Service]
Type=simple
User=ubuntu
WorkingDirectory=${APP_DIR}/api
EnvironmentFile=${APP_DIR}/api/.env
ExecStart=${APP_DIR}/api/pulsewatch-api
Restart=always
RestartSec=5

[Install]
WantedBy=multi-user.target
`;

  const webService = `[Unit]
Description=PulseWatch Web
After=network.target

[Service]
Type=simple
User=ubuntu
WorkingDirectory=${APP_DIR}/web/.next/standalone/apps/web
Environment=PORT=3000
Environment=HOSTNAME=0.0.0.0
Environment=INTERNAL_API_URL=http://127.0.0.1:4000
ExecStart=/usr/bin/node server.js
Restart=always
RestartSec=5

[Install]
WantedBy=multi-user.target
`;

  await exec(conn, `cat > /tmp/pulsewatch-api.service << 'EOF'\n${apiService}EOF`);
  await exec(conn, `cat > /tmp/pulsewatch-web.service << 'EOF'\n${webService}EOF`);
  await sudo(conn, 'mv /tmp/pulsewatch-api.service /etc/systemd/system/pulsewatch-api.service');
  await sudo(conn, 'mv /tmp/pulsewatch-web.service /etc/systemd/system/pulsewatch-web.service');

  const nginxConf = fs.readFileSync(path.join(__dirname, 'nginx', 'pulsewatch.conf'), 'utf8');
  await exec(conn, `cat > /tmp/pulsewatch-nginx << 'EOF'\n${nginxConf}EOF`);
  await sudo(conn, 'mv /tmp/pulsewatch-nginx /etc/nginx/sites-available/pulsewatch');
  await sudo(conn, 'ln -sf /etc/nginx/sites-available/pulsewatch /etc/nginx/sites-enabled/pulsewatch');
  await sudo(conn, 'rm -f /etc/nginx/sites-enabled/default');
  await sudo(conn, 'nginx -t && systemctl reload nginx');

  await sudo(conn, 'systemctl daemon-reload && systemctl enable pulsewatch-api pulsewatch-web && systemctl restart pulsewatch-api pulsewatch-web');

  await exec(conn, 'sleep 5');
  await exec(conn, 'curl -s http://127.0.0.1:4000/health');
  await exec(conn, 'curl -s -o /dev/null -w "%{http_code}" http://127.0.0.1:3000/en');
  await exec(conn, 'systemctl is-active pulsewatch-api pulsewatch-web nginx');

  conn.end();
  console.log(`\n? Deploy complete: http://${HOST}/en`);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: deploy/deploy.py
#!/usr/bin/env python3
"""Deploy PulseWatch to remote server via SSH (password auth)."""
import os
import sys
import tarfile
import io
import time
from urllib.parse import quote

try:
    import paramiko
except ImportError:
    os.system(f"{sys.executable} -m pip install paramiko -q")
    import paramiko

HOST = os.environ.get("DEPLOY_HOST")
USER = os.environ.get("DEPLOY_USER", "ubuntu")
PASSWORD = os.environ.get("DEPLOY_PASSWORD")
APP_DIR = "/opt/pulsewatch"
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

EXCLUDE = {
    "node_modules", ".next", ".git", "环境信息", ".env",
    "pulsewatch-api.exe", "__pycache__", ".cursor",
}


def run(client, cmd, sudo=False):
    if sudo:
        cmd = f"echo '{PASSWORD}' | sudo -S bash -c {repr(cmd)}"
    print(f">>> {cmd[:120]}...")
    stdin, stdout, stderr = client.exec_command(cmd, get_pty=True)
    out = stdout.read().decode(errors="replace")
    err = stderr.read().decode(errors="replace")
    code = stdout.channel.recv_exit_status()
    if out.strip():
        print(out[-2000:])
    if code != 0:
        print(f"WARN/ERR ({code}): {err[-500:]}", file=sys.stderr)
    return code, out


def make_tarball():
    buf = io.BytesIO()
    with tarfile.open(fileobj=buf, mode="w:gz") as tar:
        for dirpath, dirnames, filenames in os.walk(ROOT):
            dirnames[:] = [d for d in dirnames if d not in EXCLUDE]
            for fn in filenames:
                if fn.endswith(".exe") or fn == ".env":
                    continue
                full = os.path.join(dirpath, fn)
                rel = os.path.relpath(full, ROOT).replace("\\", "/")
                if any(x in rel for x in EXCLUDE):
                    continue
                tar.add(full, arcname=f"pulsewatch/{rel}")
    buf.seek(0)
    return buf


def main():
    print(f"Connecting to {USER}@{HOST}...")
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    client.connect(HOST, username=USER, password=PASSWORD, timeout=30)

    run(client, f"mkdir -p {APP_DIR}", sudo=True)

    # Install prerequisites
    run(client, "apt-get update -qq && apt-get install -y -qq docker.io docker-compose-plugin nginx python3 2>/dev/null || true", sudo=True)
    run(client, "systemctl enable docker nginx && systemctl start docker nginx", sudo=True)
    run(client, "usermod -aG docker ubuntu || true", sudo=True)

    # Create database
    run(client, "sudo -u postgres psql -p 6541 -tc \"SELECT 1 FROM pg_database WHERE datname='pulsewatch'\" | grep -q 1 || sudo -u postgres psql -p 6541 -c \"CREATE DATABASE pulsewatch;\"", sudo=True)

    # Upload tarball
    print("Uploading project...")
    tar_buf = make_tarball()
    sftp = client.open_sftp()
    remote_tar = "/tmp/pulsewatch-deploy.tar.gz"
    sftp.putfo(tar_buf, remote_tar)
    sftp.close()

    run(client, f"rm -rf {APP_DIR}/* && tar -xzf {remote_tar} -C / && mv /pulsewatch/* {APP_DIR}/ && rmdir /pulsewatch 2>/dev/null; rm -f {remote_tar}", sudo=True)
    run(client, f"chown -R ubuntu:ubuntu {APP_DIR}", sudo=True)

    pg_password = os.environ.get("PG_PASSWORD")
    if not pg_password and not os.environ.get("DATABASE_URL"):
        print("Set PG_PASSWORD or DATABASE_URL before deploying.", file=sys.stderr)
        sys.exit(1)
    db_url = os.environ.get("DATABASE_URL") or (
        f"postgresql://postgres:{quote(pg_password, safe='')}@127.0.0.1:6541/pulsewatch"
    )

    # Create .env
    env_cmd = f"""cat > {APP_DIR}/.env << 'ENVEOF'
DATABASE_URL={db_url}
JWT_SECRET=$(openssl rand -hex 32)
JWT_REFRESH_SECRET=$(openssl rand -hex 32)
PORT=4000
CORS_ORIGIN=http://{HOST}
NEXT_PUBLIC_API_URL=http://{HOST}:4000
SMTP_MODE=console
NODE_ENV=production
ENVEOF"""
    run(client, env_cmd.replace("$(openssl rand -hex 32)", "`openssl rand -hex 32`"))

    # Build and start
    run(client, f"cd {APP_DIR} && docker compose -f deploy/docker-compose.prod.yml --env-file .env build 2>&1 | tail -20")
    run(client, f"cd {APP_DIR} && docker compose -f deploy/docker-compose.prod.yml --env-file .env up -d")
    run(client, f"cp {APP_DIR}/deploy/nginx.conf /etc/nginx/sites-available/pulsewatch && ln -sf /etc/nginx/sites-available/pulsewatch /etc/nginx/sites-enabled/pulsewatch && rm -f /etc/nginx/sites-enabled/default && nginx -t && systemctl reload nginx", sudo=True)

    time.sleep(5)
    run(client, "curl -s http://127.0.0.1:4000/health || true")
    run(client, "curl -s -o /dev/null -w '%{http_code}' http://127.0.0.1:3000/en || true")

    client.close()
    print(f"\n✅ Deploy complete: http://{HOST}/en")


if __name__ == "__main__":
    main()
